docs/reference/c++_support.rst

```diff
@@ -0,0 +1,173 @@
+
+C++ is considered a modern programming language as of C++11. This section describes how
+HIP supports these new C++ features.
+
+C++11 support
+-------------------------------------------------------------------------------
+
+The C++11 standard introduced many new features. These features are supported in HIP
+device code, with some notable omissions. The most significant is the lack of concurrency
+support on the device. This is because the HIP device concurrency model fundamentally
+differs from the C++ model used on the host side. For example, it's not necessary or
+possible to start a new thread on the device.
+
+Certain features have restrictions and clarifications. For example, any functions using
+the ``constexpr`` qualifier or the new ``initializer lists``, ``std::move`` or
+``std::forward`` features are implicitly considered to have the ``__host__`` and
+``__device__`` execution space specifier. Also, ``constexpr`` variables that are static
+members or namespace scoped can be used from both host and device, but only for read
+access. Dereferencing a static ``constexpr`` outside its specified execution space causes
+an error.
+
+Lambdas are supported, but there are some extensions and restrictions on their usage. For
+more information, see the `Extended lambdas`_ section below.
+
+C++14 support
+-------------------------------------------------------------------------------
+
+The C++14 language features are supported, except for the updates to the memory model.
+
+C++17 support
+-------------------------------------------------------------------------------
+
+All C++17 language features are supported.
+
+C++20 support
+-------------------------------------------------------------------------------
+
+All C++20 language features are supported, but extensions and restrictions apply. C++20
+introduced coroutines and modules, which fundamentally changed how programs are written.
+HIP doesn't support these features. However, ``consteval`` functions can be called from
+host and device, even if specified for host use only.
+
+The three-way comparison operator (spaceship operator ``<=>``) works with host and device
+code.
+
+.. _language_restrictions:
+Extensions and restrictions
+===============================================================================
+
+In addition to the deviations from the standard, there are some general extensions and
+restrictions to consider.
+
+Global functions
+-------------------------------------------------------------------------------
+
+Functions that serve as an entry point for device execution are called kernels and are
+specified with the ``__global__`` qualifier. To call a kernel function, use the triple
+chevron operator: ``<<< >>>``. Kernel functions must have a ``void`` return type. These
+functions can't:
+
+* have a ``constexpr`` specifier
+* have a parameter of type ``std::initializer_list`` or ``va_list``
+* use an rvalue reference as a parameter.
+
```

Comment:
Add

'* use parameters having different sizes in host and device code, e.g. long double arguments, or structs containing long double members.'